### PR TITLE
feat: MultiStepper for the GSF

### PR DESCRIPTION
**Pull request by benjaminhuth**
Requires #1113 

This PR adds mainly the class `MultiEigenStepperLoop` with some helper code. 

It mainly adds new, independent code, the only case where it interferes with other propagation code is the stepper-concept.

The contained UnitTests cover most functionality of the `MultiEigenStepperLoop` and are built mainly as template-functions, so they can later be reused for the other MultiStepper-implementation (which will come later). The functionality is mostly tested against the `EigenStepper`.

### CODE REVIEW

**Comment by paulgessinger** [comment]
Alright, #1113 is merged. Is this ready to review now, or are additional changes needed, @benjaminhuth?

**Comment by benjaminhuth** [comment]
This should be ready to review I think!

File: Core/include/Acts/EventData/MultiComponentBoundTrackParameters.hpp
Comment (asalzburger) [inline Core/include/Acts/EventData/MultiComponentBoundTrackParameters.hpp:160]: This returns at the weight and the single component ?
If so, please indicate both in the comment

File: Core/include/Acts/EventData/MultiComponentBoundTrackParameters.hpp
Comment (asalzburger) [inline Core/include/Acts/EventData/MultiComponentBoundTrackParameters.hpp:239]: A question here, for potential future optimisation ... in the Navigation, there's often the successive call `position()`, `direction()` ... would it make sense to implement a tuple call, such that the reduction has to be done only once ?

File: Core/include/Acts/Propagator/MultiEigenStepperLoop.hpp
Comment (asalzburger) [inline Core/include/Acts/Propagator/MultiEigenStepperLoop.hpp:73]: Se my comment from before - do you have a feeling if we could gain from something like this?

File: Core/include/Acts/Propagator/MultiEigenStepperLoop.hpp
Comment (asalzburger) [inline Core/include/Acts/Propagator/MultiEigenStepperLoop.hpp:407]: Do we really need `auto` keywords here everywhere - I think these are explicit types, aren't they ?

File: Core/include/Acts/Propagator/MultiEigenStepperLoop.ipp
Comment (asalzburger) [inline Core/include/Acts/Propagator/MultiEigenStepperLoop.ipp:31]: This implicitly implies that the navigation path is a weighted mean of all paths, will this hold for all merging policies? 
(probably)

I guess the accumulatedPathLength is anyway only needed/used for the path length aborter and the path limit settings in this scope, so I *guess* this should be fine.

File: Core/include/Acts/Propagator/MultiEigenStepperLoop.ipp
Comment (asalzburger) [inline Core/include/Acts/Propagator/MultiEigenStepperLoop.ipp:48]: I don't understand the comment here - do you want to make this configurable?

File: Core/include/Acts/Propagator/MultiEigenStepperLoop.ipp
Comment (asalzburger) [inline Core/include/Acts/Propagator/MultiEigenStepperLoop.ipp:80]: Couldn't one do the pos4, mom4 merging first, then distil `pos4, dir, q/p` out of it?

Somehow I think this would be more stable, but I can wrong.

File: Core/include/Acts/Utilities/Identity.hpp
Comment (asalzburger) [inline Core/include/Acts/Utilities/Identity.hpp:14]: If `std::identity` implements the same interface, we could make this with a pragma already and alias of std::identity,  I always fear that we'll forget to remove this sort of code.

**Review by asalzburger — APPROVED** [review]
I have a view points which would be great to have some clarification on.

File: Core/include/Acts/Propagator/MultiEigenStepperLoop.hpp
Comment (asalzburger) [reply Core/include/Acts/Propagator/MultiEigenStepperLoop.hpp:407]: It's not against, it sort of hides whether the return type is e.g. a `Object&` or a `const Object&` which has a big consequence to the caller, I just wanted tot make this clear. That can be just done with a comment, then we're good, as they are accessors it's clear then.

File: Core/include/Acts/Propagator/MultiEigenStepperLoop.ipp
Comment (benjaminhuth) [reply Core/include/Acts/Propagator/MultiEigenStepperLoop.ipp:31]: Yeah, in all places where a single `pathLength`-representation for all components is needed, I did a weighted sum (e.g. for the return value of the `MultiEigenStepperLoop::step(...)`-function.
The pathlength of such a multi-component-propagation isn't really defined I think, and this is the value which comes closest to it. 
One could also use the pathlenght of the maximum-weight component e.g., but in normal use cases this should not make a huge difference.

File: Core/include/Acts/Propagator/MultiEigenStepperLoop.hpp
Comment (benjaminhuth) [reply Core/include/Acts/Propagator/MultiEigenStepperLoop.hpp:407]: Yeah I see that it might be not quite clear, that the `const`-qualified accessors are inherited from the `ComponentProxyBase`. I did this to avoid code duplications between the const and the non-const `ComponentProxy`...

I'll add a comment to clarify this.

File: Core/include/Acts/Propagator/MultiEigenStepperLoop.ipp
Comment (benjaminhuth) [reply Core/include/Acts/Propagator/MultiEigenStepperLoop.ipp:48]: That was one thing that came up in the meeting with Anthony and Christos, they said that at ATLAS they use the mode of the mixture for the final parameters, not the mean. Currently the bound states are combined by computing the mean. This probably should be tried out during the validation phase?

However, I will clarify this comment with a better formulation

File: Core/include/Acts/EventData/MultiComponentBoundTrackParameters.hpp
Comment (benjaminhuth) [reply Core/include/Acts/EventData/MultiComponentBoundTrackParameters.hpp:239]: This particular class (the `MultiComponentBoundTrackParameters<charge_t>`) should never interact with the navigation. All these accessors with the reductions are required by the `BoundTrackParametersConcept` in `TrackParametersConcept.hpp`.

However (as you also commented below), for the interaction between the `Stepper::State` and the `Navigator`, this could potentially save some computations.

File: Core/include/Acts/Propagator/MultiEigenStepperLoop.hpp
Comment (benjaminhuth) [reply Core/include/Acts/Propagator/MultiEigenStepperLoop.hpp:73]: As mentioned above: Yes this could potentially reduce both code size and increase performance here...

However, I'm not sure if I would implement this immediately or do a big performance measurement some time and keep it in mind...

File: Core/include/Acts/Propagator/MultiEigenStepperLoop.ipp
Comment (benjaminhuth) [reply Core/include/Acts/Propagator/MultiEigenStepperLoop.ipp:80]: Rigth now  I'm not sure how one would do this with the mom4... 

But now where I look at it, this function (and some others...) outputs garbage if the charges of the components are not equal, right? I will think this through...

**Comment by paulgessinger** [comment]
Will merge once the threads are resolved. @asalzburger, @benjaminhuth can you resolve the threads when complete?

File: Core/include/Acts/Propagator/MultiEigenStepperLoop.hpp
Comment (benjaminhuth) [reply Core/include/Acts/Propagator/MultiEigenStepperLoop.hpp:73]: So to resume on this:

You mean something like this, right? (taken arbitrarily from Navigator.hpp:366)

```c++
state.navigation.currentVolume = boundary->attachedVolume(
     state.geoContext, stepper.position(state.stepping),
     stepper.direction(state.stepping), state.stepping.navDir);
```

And here do something like:

```c++

const auto [pos, dir] = stepper.posdir(state.stepping);

state.navigation.currentVolume = boundary->attachedVolume(
     state.geoContext, pos, dir, state.stepping.navDir);
```

But this would change the whole stepper-navigator-interaction, right? Or do I misunderstand what your idea is?

Comment (asalzburger) [reply Core/include/Acts/Propagator/MultiEigenStepperLoop.hpp:73]: We can leave this as an optimisation step for later.

**Comment by asalzburger** [comment]
I am satisfied with the answers, we can do the optional optimisation step in a second round.

**Comment by asalzburger** [comment]
I will not add the kodiac auto-merge, as this PR is very big, it may need another rebase before merging.

**Review by asalzburger — APPROVED** [review]
Thanks for adding the 'identity' change.

**Pull request merged**
